onnxruntime/core/providers/cuda/nn/rms_norm.cc

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cuda/nn/rms_norm.h"
+#include "core/providers/cuda/nn/layer_norm.h"
+#include "core/providers/cuda/nn/layer_norm_impl.h"
+#include "core/providers/cuda/cuda_common.h"
+#include "core/providers/cpu/nn/layer_norm_helper.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+// RMSNorm uses LayerNorm kernel, which only supports X and scale both
+// being the same data type.
+#define REGISTER_KERNEL_TYPED(T)                                                              \
+  ONNX_OPERATOR_TYPED_KERNEL_EX(RMSNormalization, kOnnxDomain, 23, T, kCudaExecutionProvider, \
+                                (*KernelDefBuilder::Create())                                 \
+                                    .TypeConstraint("T", DataTypeImpl::GetTensorType<T>())    \
+                                    .TypeConstraint("V", DataTypeImpl::GetTensorType<T>()),   \
+                                RMSNorm<T, float, T, true>);
```

Comment:
For double, I suggest to use double for U (just like the simplified layer norm contrib op).